src/main/java/instructionstuff/Instruction.java

```diff
@@ -0,0 +1,288 @@
+         * @param taskList The taskList instance to add the task in.
+         * @param mainWindow The MainWindow instance to use to display messages.
+         * @throws DukeException If an error occurred during execution.
+         */
+        @Override
+        public void execute(Storage storage, TaskList taskList, MainWindow mainWindow) throws DukeException {
+            taskList.addTask(this.task);
+            String s = "Got it. I've added this task:\n " + this.task + "\nNow you have "
+                    + taskList.getSize() + " tasks in the list.";
+
+            mainWindow.setMessage(s);
+        }
+
+    }
+
+    /**
+     * A class to represent the marking of task action.
+     *
+     */
+    public static class Mark extends Instruction {
+
+        /** The index of task to mark. */
+        private int index;
+
+        /**
+         * Initialises the class fields.
+         *
+         * @param index The index of task.
+         */
+        public Mark(int index) {
+            this.index = index;
+        }
+
+        /**
+         * Marks the task at given index in given taskList.
+         *
+         * @param storage The Storage instance to store data in if needed.
+         * @param taskList The taskList instance to mark the task in.
+         * @param mainWindow The MainWindow instance to use to display messages.
+         * @throws DukeException If an error occurred during execution.
+         */
+        @Override
+        public void execute(Storage storage, TaskList taskList, MainWindow mainWindow) throws DukeException {
+            taskList.markTask(this.index);
+            assert index >= 1 && index <= taskList.getSize();
+            String s = "Nice! I've marked this task as done:\n " + taskList.getTask(index);
+            mainWindow.setMessage(s);
+        }
+    }
+
+
+    /**
+     * A class to represent the unmarking of task action.
+     *
+     */
+    public static class Unmark extends Instruction {
+
+        /**
+         * The index of task to unmark.
+         */
+        private int index;
+
+        /**
+         * Initialises the class fields.
+         *
+         * @param index The index of task to unmark.
+         */
+        public Unmark(int index) {
+            this.index = index;
+        }
+
+        /**
+         * Unmarks the task at given index in given taskList.
+         *
+         * @param storage The Storage instance to store data in if needed.
+         * @param taskList The taskList instance to unmark task in.
+         * @param mainWindow The MainWindow instance to use to display messages.
+         * @throws DukeException If an error occurred during execution.
+         */
+        @Override
+        public void execute(Storage storage, TaskList taskList, MainWindow mainWindow) throws DukeException {
+            taskList.unmarkTask(this.index);
+            assert index >= 1 && index <= taskList.getSize();
```

Comment:
I am not sure why the assertion is done before the unmarking, it's also quite interesting that the same conditions are checked within execute, and and exception is thrown.

if the index is indeed invalid, an exception is thrown in taskList.unmarkTask. Then is the assert still necessary? I am not too sure